src/smc-webapp/share/page.cjsx

```diff
@@ -103,13 +107,19 @@ CoCalcLogo = ({base_url}) ->
     <img style={height:'21px', width:'21px'} src="#{base_url}/share/cocalc-icon.svg" />
 
 TopBar_propTypes =
-    viewer     : rtypes.string
-    path       : rtypes.string
-    project_id : rtypes.string
-    base_url   : rtypes.string
-    site_name  : rtypes.string
+    viewer       : rtypes.string
+    path         : rtypes.string # The share url. Must have a leading `/`. {base_url}/share{path}
+    project_id   : rtypes.string
+    base_url     : rtypes.string
+    site_name    : rtypes.string
+    public_paths : rtypes.immutable.Map
+
+TopBar = ({viewer, path, project_id, base_url, site_name, public_paths}) ->
+    list_of_paths = public_paths
```

Comment:
@johnjeng , what if there are 1 million public paths?  This linear search through all data does not scale.  Already, there are about 30K public paths, and having to do a linear search through all of them like you've done here is not reasonable.

We need to solve this problem differently, e.g, by maintaining another data structure that is built right when the public_paths synctable is constructed, and that gets updated whenever changes come in.

Reply:
Oh, in fact, you've already built this. Changing the PR to use that.